### DISCUSSION TITLE
Generate all runtime.locale.js, but only run one.

### DISCUSSION THREAD

**Discussion by techtruth** [privatenumber/webpack-localize-assets-plugin#48] [Q&A]
I would like to only load the locale entrypoint that I need for the websites current language. In my head, this would be something like checking navigator.language and then loading the correct core/runtime file. I believe I have successfully generated all the locale core and runtimes for every language I have set in the locale json files.

As it stands now, all of the localized-entrypoints are invoking as soon as the app starts. This is likely due to the combined behavior of the HtmlWebpackPlugin that adds what seems to be all the entrypoints to a generated index.html file. In example i have a console.log(__('helloWorld')) that gets called as many times as I have locales. So 5 locales, it gets called 5 times.

In my head, a solution would be to load a 'locale bootstrap' that got the navigators language and then loaded the correct locale-backed version of the core/runtime. But not sure how practical that is for this plugin, or if HtmlWebpackPlugin would need some options.

I thought to exclude all the locale-based entrypoints and then write this bootstrapper myself, but wanted to see if there was a wheel already rolling somewhere.

I can exclude things from HtmlWebpackPlugin but it doesn't seem to want to exclude the core or runtimes. It looks like the HtmlWebpackPlugin does allow other plugins to edit the html delivered with "tappable hooks" https://github.com/jantimon/html-webpack-plugin#events But i am yet to understand how that actually works. I might know where my weekend is going.

It looks as though if webpack-localize-assets-plugin used an approach like below, it could edit/alter the html coming out to include the chunks and the bootstrap function that then loads the localized set of core/runtime to the head/body. I'll have to have a think about it a bit more before, but so far it seems doable.

**Comment by privatenumber** [comment]
Your question might be better asked on the `html-webpack-plugin` repository.

I don't use `html-webpack-plugin` but I have accomplished what you're trying to do by making each locale a Webpack entry point and loading the appropriate locale file with the application. I select the right locale file via backend but I imagine this is possible in the frontend as well as long as the app is instantiated after the appropriate locale file is loaded.

**Reply by techtruth** [reply]
I've fingerfiddled together a plugin that does not depend on html-webpack-plugin. I could see this being supported by your module, as it does not act if htmlwebpack plugin isn't installed or used in webpack. It will only act to arrange the locale data in the bootstrapper if both your plugin and htmlwebpack plugin are used. It feels to me that this could be a very nice way to inform htmlwebpackplugin that there are additional things to consider when using your plugin, that doesn't make your plugin rely on htmlwebpackplugin.

That is to say, would you be open to a PR to include this in your codebase here? I would like to get some better insights than my own if this "locale bootstrap" method is bad for reasons I am not thinking of, but aside from it being a little weird I don't see an issue... yet. :-D


```const HtmlWebpackPlugin = require("safe-require")("html-webpack-plugin");
class LocaleBootstrap {
  apply(compiler) {
    compiler.hooks.compilation.tap("LocaleBootstrap", compilation => {
      //Intercept localized assets and remove them from the asset queue.
      // Construct a new js asset which determines browser locale at runtime
      // and imports the matching locale bundle.
      HtmlWebpackPlugin.getHooks(compilation).beforeAssetTagGeneration.tapAsync(
        "LocaleBootstrap",
        (data, cb) => {
          //Get locale assets
          let localeAssets = data.assets.js.filter(asset =>
            asset.match("/locale/")
          );
          
          //Create the locale bootstrap file
          let fileString = \`let clientLocale = navigator.language;\n\`;
          fileString += `console.debug("Localized to", clientLocale);\n`;
          fileString += `let localeArray = new Array(\n`;
          for (let asset of localeAssets) {
            fileString += `"${asset}",`;
          } 
          fileString = fileString.replace(/,$/, ""); //Remove trailing comma
          fileString += `);\n`; //End of array
          fileString += `let neededAssets = localeArray.filter( asset =>
                           asset.match("/locale/"+clientLocale+"/")
                         );\n`;

          fileString += `for( asset of neededAssets ) {
                              import(asset);
                         }`;
          fileString += `console.debug("Localization complete.")`;
          
          compilation.emitAsset(
            "./js/localebootstrap.js",
            new compiler.webpack.sources.RawSource(fileString)
          );
          
          //Remove the localeAssets from tags to include in html
          // (because they will be loaded from locale bootstrap file)
          data.assets.js = data.assets.js.filter(
            asset => !asset.match("/locale/")
          );
          //Add the created asset to the files
          data.assets.js.push("./js/localebootstrap.js");
          
          // Tell webpack to move on
          cb(null, data);
        } 
      );
    });
  } 
} 

module.exports = LocaleBootstrap;
```
